File: solutions/ts/24.ts
type Tree = { value: string; left?: Tree; right?: Tree } | undefined;

function isTreesSynchronized(tree1: Tree, tree2: Tree): [boolean, string] {
  function validateSync(node1: Tree, node2: Tree): boolean {
    let subSync1 = true;
    let subSync2 = true;
    if (node1?.left && node2?.right) {
      subSync1 = validateSync(node1.left, node2.right);
    }
    if (node1?.right && node2?.left) {
      subSync2 = validateSync(node1.right, node2.left);
    }
    const sameRoot = node1?.value === node2?.value;
    return sameRoot && subSync1 && subSync2;
  }
  const sync = validateSync(tree1, tree2);
  const firstRoot = tree1!.value;
  return [sync, firstRoot];
}

const baseTree1 = {
  value: "🎄",
  left: { value: "⭐" },
  right: { value: "🎅" },
};

const baseTree2 = {
  value: "🎄",
  left: { value: "🎅" },
  right: { value: "⭐" },
};

const sync1 = isTreesSynchronized(baseTree1, baseTree2); // [true, '🎄']
console.log(sync1);

/*
  tree1          tree2
   🎄              🎄
  / \             / \
⭐   🎅         🎅   ⭐
*/

const baseTree3 = {
  value: "🎄",
  left: { value: "🎅" },
  right: { value: "🎁" },
};

const sync2 = isTreesSynchronized(baseTree1, baseTree3); // [false, '🎄']
console.log(sync2);

const baseTree4 = {
  value: "🎄",
  left: { value: "⭐" },
  right: { value: "🎅" },
};

const sync3 = isTreesSynchronized(baseTree1, baseTree4); // [false, '🎄']
console.log(sync3);

const sync4 = isTreesSynchronized({ value: "🎅" }, { value: "🧑‍🎄" }); // [false, '🎅']
console.log(sync4);

/* 
  Score: ★★★★★★★★ (3 bonus stars for solving it today!)
  Points: 180
  ops: 3144
*/

File: solutions/python/24.py
def is_trees_synchronized(tree1, tree2):
    def validate_sync(node1, node2):
        sub_sync_1 = True
        sub_sync_2 = True
        if (node1.get("left") and node2.get("right")):
            sub_sync_1 = validate_sync(node1.get("left"), node2.get("right"))
        if (node1.get("right") and node2.get("left")):
            sub_sync_2 = validate_sync(node1.get("right"), node2.get("left"))
        same_root = node1.get("value") == node2.get("value")
        return same_root and sub_sync_1 and sub_sync_2

    sync = validate_sync(tree1, tree2)
    first_root = tree1["value"]
    return [sync, first_root]


tree1 = {
    "value": "🎄",
    "left": {"value": "⭐"},
    "right": {"value": "🎅"},
}

tree2 = {
    "value": "🎄",
    "left": {"value": "🎅"},
    "right": {"value": "⭐"},
}

sync1 = is_trees_synchronized(tree1, tree2)  # [true, '🎄']
print(sync1)

"""
  tree1          tree2
   🎄              🎄
  / \             / \
⭐   🎅         🎅   ⭐
"""

tree3 = {
    "value": "🎄",
    "left": {"value": "🎅"},
    "right": {"value": "🎁"},
}

sync2 = is_trees_synchronized(tree1, tree3)  # [false, '🎄']
print(sync2)

tree4 = {
    "value": "🎄",
    "left": {"value": "⭐"},
    "right": {"value": "🎅"},
}

sync3 = is_trees_synchronized(tree1, tree4)  # [false, '🎄']
print(sync3)

sync4 = is_trees_synchronized({"value": "🎅"}, {"value": "🧑‍🎄"})  # [false, '🎅']
print(sync4)


"""
  Score: ★★★★★★★ (3 bonus stars for solving it today!)
  Points: 180
  ops: 3333
"""
